### ISSUE TITLE
Large number of node_modules/files causes "too many open files" on OS X

### ISSUE THREAD

**Issue by emlai** [meteor/meteor#6952]
**_1 Upvote**_ I have ~40000 files (~7GB altogether) in my `public` directory. When I run `meteor`, the "building the application" step takes 3 minutes. When I then press <kbd>ctrl</kbd><kbd>c</kbd> to exit it, the following gets printed:

```
fs.js:439
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^
Error: EMFILE, too many open files '/Users/emlai/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/static-assets/server/shell-server.js'
    at Object.fs.openSync (fs.js:439:18)
    at Object.fs.readFileSync (fs.js:290:15)
    at Object.Module._extensions..js (module.js:473:44)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at /tools/runners/run-app.js:146:7
    at /Users/emlai/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/tool-env/cleanup.js:24:7
    at Array.forEach (native)
    at Function._.each._.forEach (/Users/emlai/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
    at /Users/emlai/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/tool-env/cleanup.js:23:7
    at Object.exports.noYieldsAllowed (/tools/utils/fiber-helpers.js:37:14)
    at runHandlers (/Users/emlai/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/tool-env/cleanup.js:20:16)
    at process.<anonymous> (/Users/emlai/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/tool-env/cleanup.js:32:5)
    at process.g (events.js:180:16)
    at process.emit (events.js:92:17)
    at Signal.wrap.onsignal (node.js:800:46)
```

My setup: Meteor 1.3.2.4 on OS X El Capitan 10.11.5 Beta.

```
» meteor list
accounts-facebook      1.0.9  Login service for Facebook accounts
accounts-password      1.1.8  Password support for accounts
accounts-ui            1.1.9  Simple templates to add login widgets to an app
blaze-html-templates   1.0.4  Compile HTML templates into reactive UI with Meteor Blaze
ecmascript             0.4.3  Compiler plugin that supports ES2015+ in all .js files
es5-shim               4.5.10  Shims and polyfills to improve ECMAScript 5 support
http                   1.1.5  Make HTTP calls to remote servers
jquery                 1.11.8  Manipulate the DOM using CSS selectors
meteor-base            1.0.4  Packages that every Meteor app needs
mobile-experience      1.0.4  Packages for a great mobile user experience
mongo                  1.1.7  Adaptor for using MongoDB and Minimongo over DDP
reactive-dict          1.1.7  Reactive dictionary
reactive-var           1.0.9  Reactive variable
sergeyt:typeahead      0.11.1_8  Autocomplete package for meteor powered by twitter typ...
standard-minifier-css  1.0.6  Standard css minifier used with Meteor apps by default.
standard-minifier-js   1.0.6  Standard javascript minifiers used with Meteor apps by de...
tracker                1.0.13  Dependency tracker to allow reactive callbacks
```

**Comment by abernix** [comment]
@emlai What's the structure of the files in your `public` directory?  Just 40k files in the directory?  Or are they organized into sub-directories?

What is the output of running `launchctl limit maxfiles` on your terminal?  It should show what the user-limits are for your account.

**Comment by emlai** [comment]
@abernix They're organized into ~200 subdirectories inside `public`. Inside those subdirectories, there are 1-3 directories, inside which the actual files are located.

```
» launchctl limit maxfiles
	maxfiles    256            unlimited  
```

**Comment by abernix** [comment]
@emlai Ok, that's pretty reasonably matrixing of your files.  It's showing your soft limit is `256` but your hard limit is `unlimited`, however, I think `unlimited` by default is still only 10k files on El Cap.  Out of curiosity, what happens if you raise your user limits *waaaay* up (temporarily):

`sudo launchctl limit maxfiles 524288 524288`

Then try Meteor again.

Once you get the results (good or bad), you can put them back to how they were with:

`sudo launchctl limit maxfiles 256 unlimited`

(These are normally controlled by `launchd` at boot time, so absolute worse case a reboot will put you back exactly as things were before)

**Comment by emlai** [comment]
Using `sudo launchctl limit maxfiles 524288 524288` didn't change anything.

**Comment by abernix** [comment]
Ok, the specifics of actually changing the file limit are a bit hairy in OS X but I can confirm that this breaks for me too.  It's highly possible that this would be fixed if Meteor used `graceful-fs` instead of `fs` as I'm sure openSync is, well, not being graceful, but not sure if that will take care of Windows, or if putting 40000 files into the `public/` directory is something permitted. :grin:

**Reproduction Repo**: https://github.com/abernix/bunch-o-files

Basically, this is a new meteor project with this commands:
```
for i in {1..40000}; do openssl rand -base64 16 | md5 | head -c16 | sed -e 's/^\(\(...\)\(.\).*\)$/mkdir -p public\/\2\/\3\/ \&\& touch public\/\2\/\3\/\1/'; done | sh
```

**Comment by queso** [comment]
I don't think this issue is related to just `public/`. I am getting this error in an app that has a lot of node_module dependencies from Chimp and a fair amount of files in the package system too. The project is based on Rocket Chat if that helps at all.

**Comment by queso** [comment]
So I killed off my `.meteor/local` folder and tried to run the app again without the Chimp npm dep installed. I got this stacktrace as a result:

````
W20160504-22:35:29.512(-4)? (STDERR)     at Object.fs.readFileSync (fs.js:290:15)
W20160504-22:35:29.513(-4)? (STDERR)     at Object.Module._extensions..json (module.js:480:44)
W20160504-22:35:29.513(-4)? (STDERR)     at Module.load (module.js:356:32)
W20160504-22:35:29.513(-4)? (STDERR)     at Function.Module._load (module.js:312:12)
W20160504-22:35:29.513(-4)? (STDERR)     at Module.require (module.js:364:17)
W20160504-22:35:29.514(-4)? (STDERR)     at require (module.js:380:17)
W20160504-22:35:29.514(-4)? (STDERR)     at Object.<anonymous> (/Users/josh/.meteor/packages/webapp/.1.2.8.110qi9z++os+web.browser+web.cordova/npm/node_modules/send/node_modules/mime/mime.js:87:13)
W20160504-22:35:29.514(-4)? (STDERR)     at Module._compile (module.js:456:26)
W20160504-22:35:29.514(-4)? (STDERR)     at Object.Module._extensions..js (module.js:474:10)

Error: UNKNOWN, open '/Users/josh/Code/Clients/secretapp/packages/rocketchat-statistics/server/functions/getAverage.coffee'
```

**Comment by focused** [comment]
To reproduce this bug you can try meteorchief base repo https://github.com/themeteorchef/base
Just clone, install packages and run, nothing happens at `localhost:3000` in browser, and when hit Ctrl+C:
```
fs.js:439
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^
Error: EMFILE, too many open files '/Users/focused/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/static-assets/server/shell-server.js'
    at Object.fs.openSync (fs.js:439:18)
    at Object.fs.readFileSync (fs.js:290:15)
    at Object.Module._extensions..js (module.js:473:44)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at /tools/runners/run-app.js:146:7
    at /Users/focused/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/tool-env/cleanup.js:24:7
    at Array.forEach (native)
    at Function._.each._.forEach (/Users/focused/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
    at /Users/focused/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/tool-env/cleanup.js:23:7
    at Object.exports.noYieldsAllowed (/tools/utils/fiber-helpers.js:37:14)
    at runHandlers (/Users/focused/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/tool-env/cleanup.js:20:16)
    at process.<anonymous> (/Users/focused/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/tool-env/cleanup.js:32:5)
    at process.g (events.js:180:16)
    at process.emit (events.js:92:17)
    at Signal.wrap.onsignal (node.js:800:46)
```
Mac OS El Captain 10.11.4, node 0.10.44, npm 2.15.0

**Comment by mateodelnorte** [comment]
👍

**Comment by queso** [comment]
@focused: I actually tried and couldn't reproduce this by downloading base... I cloned the repo, did a `meteor npm i`, then did a `meteor run`. Compiled and ran fine on my MacBook 12".

**Comment by mateodelnorte** [comment]
This is happening to 5 members on my team, and began immediately after the 1.3 upgrade. We don't have nearly that many files our public directory (maybe 30 total), but we do now have a large amount of files nested under ./node_modules/.

**Comment by merlinpatt** [comment]
What's the purpose of these 40000 files?

**Comment by focused** [comment]
@queso What is your OSX, npm, node and Meteor versions? What if you enter `ulimit -n` and `launchctl limit maxfiles`?

On my mac pro:
```
focused:base focused$ launchctl limit maxfiles
	maxfiles    256            unlimited
focused:base focused$ ulimit -n
1024
focused:base focused$ ulimit
unlimited
```

**Comment by abernix** [comment]
Those experiencing this, try to update this issue with the output of this command, run from your project root.  It will print the top 5 directories with the most files and should be super discreet.  It's generally expected that your `.meteor` directory will have 50k+ but the others might be causing problems.  (This will only work for *nix users).  Just for some extra possibly-insightful stats on this.
```
find . -type f | cut -d / -f 2 | sort | uniq -c | sort -rn | head -5
```

**Comment by focused** [comment]
```
focused:base focused$ find . -type f | cut -d / -f 2 | sort | uniq -c | sort -rn | head -5
60831 node_modules
 609 .meteor
  34 imports
  22 .git
  10 client
focused:base focused$
```

**Comment by focused** [comment]
Today I added Semantic-UI package https://github.com/Semantic-Org/Semantic-UI-Meteor to a small meteor project and... the same error!
```
=> App running at: http://localhost:3000/
^C
fs.js:439
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^
Error: EMFILE, too many open files '/Users/focused/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/tools/static-assets/server/shell-server.js'
    at Object.fs.openSync (fs.js:439:18)

...
```

**Comment by focused** [comment]
Please help! :(

When my project grows at some moment i find myself stuck, i tried to increase all possiible limits on OSX, but meteor fails to start! I tried different packages, but after adding 5th or 6th meteor or npm package, it happens again! I completely remove node, nvm and reinstall meteor.

I even start to use only npm integrated into meteor: `meteor npm install ...`

How can i try other versions of meteor? Maybe downgrading to Meteor 1.3.0 solve the issue???

**Comment by focused** [comment]
`meteor run --release 1.3.1` does not help

**Comment by focused** [comment]
Repeat with my friend all steps on freshly installed Ubuntu 14.04, everything works, so it's only OSX issue.

**Comment by focused** [comment]
Finally solve the problem by creating new user in OS and reinstall meteor.

**Comment by abernix** [comment]
I'd be surprised if this situation went away completely with a new user, unless your old user had a `~/.bashrc` or `~/.bash_history` line which was setting your max_files `ulimit` particularly low.  (very possible though, as this setting used to be very common!).  `ulimit -n` will show your setting for the current shell session.

Things will definitely work better on Ubuntu which has a much higher default `maxfiles` settings than OS X does, which is by default (I believe, set unreasonably low) at 256 files.  It's possible a new user may have higher defaults these days, but I haven't created a new OS X user in ages, so I can't say for sure.  It may happen that it will happen again as you add more files, but please feel free to try my [40,000 file reproduction repo](https://github.com/abernix/bunch-o-files) and see if it works for you – it would be great if it did!

Meteor tries to make things better by [setting your open-files `ulimit` as high as allowed](https://github.com/meteor/meteor/blob/965ad94df3e11a77530310b1b46c9dfd25835d0a/meteor#L117-L127) but unfortunately, starting with OS X Yosemite, it became much more difficult to change `ulimit` "maximum" settings and it now involves changes to LaunchDaemon plists, etc.  In fact, if you have it set manually to a number higher than 16384, Meteor will actually *lower* your maxfiles ulimit. (Though this part could easily be fixed).

Regardless, even if you could get Mac's `ulimit -n` higher, things will be slow with watching that many files.  Performance could probably be improved if Meteor switched to using the more efficient `fs.watch` instead of `fs.watchFile`, which is what it uses right now.  With the addition of watches on `node_modules` in 1.3 (and thus a much larger number of files), it might also be beneficial if Meteor tried using third-party file-watching solutions, perhaps the NPM packages [`chokidar`](https://github.com/paulmillr/chokidar) or [`graceful-fs`](https://github.com/isaacs/node-graceful-fs), [FB's `watchman`](https://github.com/facebook/watchman) or some combination thereof.

**Comment by focused** [comment]
I tried your repo with new MacOS user:
```
owls:bunch-o-files owl$ meteor npm install
...
owls:bunch-o-files owl$ meteor -p 4000
[[[[[ ~/Development/meteor/bunch-o-files ]]]]]

=> Started proxy.
=> Started MongoDB.
=> Started your app.

=> App running at: http://localhost:4000/

/Users/owl/.meteor/packages/templating/.1.1.9.1h5n3gh++os+web.browser+web.cordova/plugin.compileTemplatesBatch.os/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:116
      throw error;
            ^
Error: UNKNOWN, open '/Users/owl/Development/meteor/bunch-o-files/public/32e/0/32e0493e40bb458a'
owls:bunch-o-files owl$
```

**Comment by abernix** [comment]
@focused Thanks for the results.  Good to know.

**Comment by pyry** [comment]
I had the same issue with Meteor 1.3.2.4 on OS X Yosemite 10.10.5 and running several instances at the same time (tests, react storybook and meteor itself):

```
find . -type f | cut -d / -f 2 | sort | uniq -c | sort -rn | head -5
122642 node_modules
 980 .meteor
  81 public
  60 imports
   4 client
```

however for me `sudo launchctl limit maxfiles 524288 524288` seemed to solve the issue!

**Comment by abernix** [comment]
@pyry What does `ulimit -n` (without sudo) say for your system?

**Comment by pyry** [comment]
@abernix 
```
ulimit -n
2560
```

**Comment by abernix** [comment]
Maybe `node_modules` and non-`node_modules` are watched differently then because if this fixed it for you, I find it strange [because this code explicitly lowers your ulimit](https://github.com/meteor/meteor/blob/965ad94df3e11a77530310b1b46c9dfd25835d0a/meteor#L117-L127) for the current process.

Only running `sudo launchctl limit` locally didn't work for me, but if anyone else is willing to try [from a Meteor checkout](https://github.com/meteor/meteor/blob/release-1.3.3/README.md#slow-start-for-developers), on OS X 10.10.x and 10.11.x I have personally only been able to **improve** this problem by:

* Commenting out the Meteor [code that changes the ulimit](https://github.com/meteor/meteor/blob/965ad94df3e11a77530310b1b46c9dfd25835d0a/meteor#L117-L127) (same as linked above)
* Raising my maxfiles up using [this launchctl plist](https://gist.github.com/abernix/854c70cf382957af3475e4e044e73f38#file-limit-maxfiles-plist) – though you should consult [other instructions for better details](http://docs.basho.com/riak/kv/2.1.4/using/performance/open-files-limit/#adjusting-open-file-limits-in-yosemite).  **Make sure you reboot.**
* Make sure you do not have *anything* else setting your ulimit in your `.bashrc` or `.bash_profile` (or other shell config) as `ulimit -n` can only be set once per process (or only lowered thereafter).
* If everything is correct, when you type `ulimit -n` it should print whatever value you have in that plist.

Meteor keeps spinning `\ | / | \ |` ... for a very long time after the app is initiated, but it seems to eventually stabilize (though I did still have a random seg-fault).  Changes and reloads are still a bit slow, but they seem to work.

For now, I would still generally recommend avoiding having huge node dependency trees or keeping massive numbers of files around your project directory.  Use Amazon S3 or similar to store excessive numbers of files where possible.  Watching this many files (and thus providing the hot reloading that Meteor uses) will certainly be a bit sluggish.  Once upgrades to Node are done, `fs.watchFiles` (and `fs.watch`) should improve a bit.

**Comment by PEM--** [comment]
Nasty bug. This settings did the trick for me: [How to persist ulimit on OSX](http://unix.stackexchange.com/questions/108174/how-to-persist-ulimit-settings-in-osx-mavericks).

**Comment by steve-ross** [comment]
Since upgrading to 1.4.2.1 today I'm getting an error :( 

my project:
```
13980 .meteor
13481 node_modules
5301 .git
 520 client
 145 server
```

**Comment by abernix** [comment]
@steve-ross Meteor 1.4.2.1 has a lot of performance enhancements which make it significantly faster and, but file watching is more important than ever.  Meteor tries to use as many file watchers as it needs, but unfortunately, the default file watching limits on OS X are relatively limited and these cannot be raised programmatically – a change to your system might be the only way to resolve this.  What is the output of the following command on your system?

    sysctl -n kern.maxfilesperproc

**Comment by steve-ross** [comment]
@abernix Unfortunately I've followed every guide out there.  I spent the whole day yesterday trying to fix it and have resorted to installing a VM or I have to run meteor as sudo was the only way to change the ulimit. You guys dev/null the ulimit switch:  so no one actually knows whether or not the ulimit actually gets changed.

here is where ulimit gets dev/null-ed:
```
if [ "$(ulimit -n)" != "unlimited" ] && ! [ "$(ulimit -n)" -gt 32768 ] ; then
    ulimit -n 32768 > /dev/null 2>&1 || \
    ulimit -n 16384 > /dev/null 2>&1 || \
    ulimit -n 8192 > /dev/null 2>&1 || \
    ulimit -n 4096 > /dev/null 2>&1 || \
    ulimit -n 2048 > /dev/null 2>&1 || \
    ulimit -n 1024 > /dev/null 2>&1 || \
    ulimit -n 512 > /dev/null 2>&1
fi
```

Screencast of me trying to change my ulimit:
![screen capture on 2016-11-11 at 10-23-05](https://cloud.githubusercontent.com/assets/297351/20219925/7dcd85aa-a7f9-11e6-9620-ec5f5d81273e.gif)

**Comment by hwillson** [comment]
@steve-ross Try the following (will work with OS X El Capitan):

```
$ echo kern.maxfiles=65536 | sudo tee -a /etc/sysctl.conf
$ echo kern.maxfilesperproc=65536 | sudo tee -a /etc/sysctl.conf
$ sudo sysctl -w kern.maxfiles=65536
$ sudo sysctl -w kern.maxfilesperproc=65536
$ ulimit -n 65536 65536 
```

**Comment by steve-ross** [comment]
@hwillson ``` ulimit -n 65536 65536``` fails with too many arguments.
``` ulimit -n 65536``` fails with:
ulimit: Permission denied when changing resource of type 'Maximum number of open file descriptors'

Also... my screencast basically showed what you sent not working. What do you see on your box when you run ```ulimit -a``` after you ran those commands because although mine succeed they don't actually change them.

Looks like i need to go back to before 1.4.2.x

**Comment by hwillson** [comment]
> ulimit -n 65536 65536 fails with too many arguments.

That's strange - that should definitely work? 

![screenshot 2016-11-11 10 48 53](https://cloud.githubusercontent.com/assets/137740/20220691/7cf28ee8-a7fc-11e6-97f2-d8c19e05f439.png)

After running `ulimit -a`:

![screenshot 2016-11-11 10 49 44](https://cloud.githubusercontent.com/assets/137740/20220723/9ab44a3e-a7fc-11e6-982e-f6a75193331a.png)

**Comment by abernix** [comment]
@hwillson I've definitely had issues with those commands in my days as well because of conflicts with other `ulimit` commands that had already been run.  I believe the situation in this case is that it's already been set once.

@steve-ross Thanks for providing that.  First, let me assure you that this is a complicated and unclear situation thanks to some OS X changes over the years.  Is there any chance of you providing the output of the command I asked for as well?  It's actually different than any of the things you showed.  There are a lot of problems here actually:

* First of all, the code snippet you pasted above is not from Meteor 1.4.2.x – not sure where you found it?  The current launcher is actually much intelligent.
* The way you are switching to root and changing `ulimit` and changing back to your normal user will never work.  The `ulimit` can only be raised ONCE per shell session and when you change to root, you start a new shell, and when you leave it, you resume your usual shell (whatever it was before).  Also, many times users have settings in their startup RC files which (try to) set the `ulimit` "high" but this is actually problematic as it can only be raised once per session and only lowered after that (even if it might have been possible originally!)
* I'm really not sure `/etc/sysctl.conf` file works in any way on OS X El Capitan (or from Yosemite on).  The `/Library/LaunchDaemons/limit.maxfiles.plist` is the option.  Please read [this page](https://docs.basho.com/riak/kv/2.1.4/using/performance/open-files-limit/#adjusting-open-file-limits-in-yosemite) and remove anything that is no longer relevant on your version of OS X.
* When you run `launchctl limit`, you can see that your OS is still setting a soft limit (one which you cannot go above as a regular user – ever)  5120 files.

Just for reference, here are my settings from my El Capitan machine:

```sh
$ cat /Library/LaunchDaemons/limit.maxproc.plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple/DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
  <plist version="1.0">
    <dict>
      <key>Label</key>
        <string>limit.maxproc</string>
      <key>ProgramArguments</key>
        <array>
          <string>launchctl</string>
          <string>limit</string>
          <string>maxproc</string>
          <string>2048</string>
          <string>2048</string>
        </array>
      <key>RunAtLoad</key>
        <true />
      <key>ServiceIPC</key>
        <false />
    </dict>
  </plist>
$ cat /Library/LaunchDaemons/limit.maxfiles.plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
  <plist version="1.0">
    <dict>
      <key>Label</key>
        <string>limit.maxfiles</string>
      <key>ProgramArguments</key>
        <array>
          <string>launchctl</string>
          <string>limit</string>
          <string>maxfiles</string>
          <string>524288</string>
          <string>524288</string>
        </array>
      <key>RunAtLoad</key>
        <true/>
      <key>ServiceIPC</key>
        <false/>
    </dict>
  </plist>
$ sysctl -n kern.maxfilesperproc # this is the most important value
524288
```

Make sure you have nothing in your `.zshrc` or `.bashrc` that is setting your ulimit.  I would also reboot after making changes to be sure.  But the most important command on OS X El Capitan is the `sysctl -n kern.maxfilesperproc` output.

**Comment by steve-ross** [comment]
I appreciate that it's difficult like I said... I literally spent the whole day on this yesterday so I've done all this. 

```
sysctl -n kern.maxfilesperproc                                                                                                        normal
65536

at /Library/LaunchDaemons/limit.maxfiles.plist                                                                                       normal
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
  <plist version="1.0">
    <dict>
      <key>Label</key>
        <string>limit.maxfiles</string>
      <key>ProgramArguments</key>
        <array>
          <string>launchctl</string>
          <string>limit</string>
          <string>maxfiles</string>
          <string>65532</string>
          <string>65532</string>
        </array>
      <key>RunAtLoad</key>
        <true/>
      <key>ServiceIPC</key>
        <false/>
    </dict>
  </plist>

cat /Library/LaunchDaemons/limit.maxproc.plist                                                                                        normal
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple/DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
  <plist version="1.0">
    <dict>
      <key>Label</key>
        <string>limit.maxproc</string>
      <key>ProgramArguments</key>
        <array>
          <string>launchctl</string>
          <string>limit</string>
          <string>maxproc</string>
          <string>2048</string>
          <string>2048</string>
        </array>
      <key>RunAtLoad</key>
        <true />
      <key>ServiceIPC</key>
        <false />
    </dict>
  </plist>

```

**Comment by steve-ross** [comment]
navigated to that command via the last commit message in this ticket regarding changing the ulimit... however it is still dev/nulled (see below) so it would be hard to tell if/when it doesn't work:

https://github.com/meteor/meteor/blob/340f0f6fd416623489d7cd904402f9c07b3c8df3/meteor#L125

**Comment by steve-ross** [comment]
Just a little more background: I've even gone so far as to create a new user account and see if it was my ~/.bashrc or ~/.bash_proflie  From speaking to another dev that I know he is also having issues on Sierra (he doesn't do meteor) and has spent quite a lot of time doing some of the same. Seems all Sierra has done is bumped the limit up a bit by default on the last release.

**Comment by abernix** [comment]
@steve-ross On your existing user, what happens if you open a completely brand new terminal shell and run?:

    ulimit -Sn 65536

Also, I didn't include the output of mine earlier, but for the sake of completeness:

```sh
$ launchctl limit maxfiles
	maxfiles    524288         524288
$ sysctl -a | grep maxfiles
kern.maxfiles: 524288
kern.maxfilesperproc: 524288
```

Other things I want to point out, in your case, right now are:

* The `/etc/sysctl.conf` values don't seem to matter anywhere, though I'm not surprised by this as it is not supposed to be used anymore.
* I think you should try much higher values, for example `5120` is lower than your number of files in your project.

**Comment by steve-ross** [comment]
@abernix what point release are you on? Think this was a recent 'fix' by Apple

output: : ( 
```
♆ > ulimit -Sn 65536
-bash: ulimit: open files: cannot modify limit: Invalid argument
♆ > launchctl limit maxfiles
	maxfiles    65532          65532
```

**Comment by abernix** [comment]
@steve-ross: And what if you run?:

    ulimit -Sn 256

**Comment by steve-ross** [comment]
@abernix That works...

**Comment by steve-ross** [comment]
@abernix so I was able to change it by doing the following but man this is a pain in the ass... makes me wonder if I have some leftover crap config files and I need a fresh install.

```sh
sudo -i
launchctl limit maxfiles 510000 510000
ulimit -Hn 510000
ulimit -Sn 510000
su sross
```

**Comment by steve-ross** [comment]
HOLY SHIT. 
```
♆ > cat /etc/profile
# System-wide .profile for sh(1)

if [ -x /usr/libexec/path_helper ]; then
    eval `/usr/libexec/path_helper -s`
fi

if [ "${BASH-no}" != "no" ]; then
    [ -r /etc/bashrc ] && . /etc/bashrc
fi
ulimit -n 4096
```

thx for all the help, the plist files are working now... gah!

**Comment by benjamn** [comment]
Back at Facebook that's what we called a "clown shoe," meaning a bug that's really obvious when you stumble across it, but nobody can say quite how it got there.

**Comment by steve-ross** [comment]
@benjamn stealing that

**Comment by abernix** [comment]
@steve-ross Haha.  Glad to hear you got that sorted out!

If this turns out to be a recurring problem, some improvements to the `meteor` start-up script could be made to do some reconnaissance in search of mistakes that may be limiting the ability to raise the `ulimit`.   For example, `grep`-ing various problem areas (`/etc/profile`, `~/{.bashrc,.zshrc,...}`, etc.) upon `ulimit` failure.  (Obviously this process should be named after some clown-themed horror movie.)

**Comment by LDubya** [comment]
Having this issue on OS X Sierra. Can't solve it, have been trying for the past 2 days. It first came up when I ran Meteor Update to the latest version, 1.4.2.3. Previously it was at 1.3.4.3.

I checked bashrc, zshrc, etc, nothing there effecting this
I added the limit.maxfiles.plist (and restarted the computer), and I still get the error when I run `meteor`

As part of the build process, I package the app for electron using the meteor-electron from GitHub

```
I20170113-18:40:49.226(-5)? ATOM: [warn] kqueue: Too many open files
I20170113-18:40:49.226(-5)?
I20170113-18:40:49.227(-5)? ATOM: [warn] kqueue: Too many open files
I20170113-18:40:49.227(-5)? [err] evsignal_init: socketpair: Too many open files
I20170113-18:40:49.228(-5)? [err] evsignal_init: socketpair: Too many open files
```

**Comment by LDubya** [comment]
Hey guys, it's me again. I still am unable to fix this. I've tried everything. I don't even have that many files. Please help.

```
find . -type f | cut -d / -f 2 | sort | uniq -c | sort -rn | head -5

17342 private
8756 .meteor
3126 node_modules
 836 .git
 391 .meteor-electron
```

```
ulimit -n
524288
```

On OS X Sierra. After trying all reasonable limits, I set the max file size as high as possible to see if that would work and that's where I am now. No matter what number I set (and restart the computer) it doesn't stop the console from erring out too many open files. It is set inside the limit.maxfiles.plist file. It's not set anywhere else (not in the profile, bashrc, zshrc, etc). 

Someone please help =(

**Comment by LDubya** [comment]
Bump. Help? Anyone?

**Comment by abernix** [comment]
@LDubya Does the problem go away if you get rid of the 17342 files in your `private/` folder?

**Comment by LDubya** [comment]
@abernix I still get too many open files.

```
I20170217-13:14:05.068(-5)? ATOM: [warn] kqueue: Too many open files
I20170217-13:14:05.069(-5)? [err] evsignal_init: socketpair: Too many open files
```

and if I list my file count...

```
find . -type f | cut -d / -f 2 | sort | uniq -c | sort -rn | head -5

8969 .meteor
2778 node_modules
 207 .meteor-electron
 112 public
  37 client
```

any ideas? I have never been able to get this to work and it's wearing me out.

**Comment by benjamn** [comment]
@LDubya You can disable one of the primary sources of open files by forcing the file watcher to use polling instead of OS file change notifications:
```sh
export METEOR_WATCH_FORCE_POLLING=1
meteor run
```
Note that this will cause the file watcher to `stat` your files every 500ms, which may increase idle CPU usage, but that is probably less of a problem than `EMFILE` errors.

If 500ms is too frequent (or not frequent enough!), you can set the polling interval explicitly:
```sh
export METEOR_WATCH_POLLING_INTERVAL_MS=1000
```

**Comment by LDubya** [comment]
@benjamn this seems to be allowing meteor to run.

If I wanted to disable it, is it as easy as:?

```bash
export METEOR_WATCH_FORCE_POLLING=0
```

**Comment by benjamn** [comment]
That's right, or `unset METEOR_WATCH_FORCE_POLLING`. Note that environment variables only apply to the current shell/terminal window, so you might need to put the `export` in a `.bashrc` file (depending on your shell) if you want it to be set in every new shell you create.

**Comment by keyscores** [comment]
+1 export METEOR_WATCH_FORCE_POLLING=1 solved it for me. @benjamn are you recommending this be the default setting for OSX?

**Comment by partounian** [comment]
@LDubya That link will show you how to create plist files that make the file limit values stick, I used different numbers but the point is still useful of file format and location.
https://www.chrissearle.org/2016/10/01/too-many-open-files-on-osx-macos/

**Comment by matmar10** [comment]
+1 `export METEOR_WATCH_FORCE_POLLING=1` also solved it for me. Thanks @benjamn

**Comment by abernix** [comment]
We believe this problem should be substantially improved by changes in Meteor 1.5.1 which drastically reduce the number of "watched" files and thus the number of concurrently open file descriptors.  While this change also means that you may have to wait up to 5 seconds for the first change to be detected on a particular file, further changes should be quite quick once the watcher is created.  To quote [the `History.md` from Meteor 1.5.1](https://github.com/meteor/meteor/blob/devel/History.md#v151-2017-07-12):

> open file descriptors will now be bounded roughly by the number of files you are actively editing, rather than the number of files involved in the build (often thousands)

See also: https://github.com/meteor/meteor/pull/8866.

**Comment by molerat619** [comment]
I've run these commands:

echo kern.maxfiles=65536 | sudo tee -a /etc/sysctl.conf
echo kern.maxfilesperproc=65536 | sudo tee -a /etc/sysctl.conf
sudo sysctl -w kern.maxfiles=65536
sudo sysctl -w kern.maxfilesperproc=65536
ulimit -n 65536

And since then my mysql database in my other nodejs application stopped working. I opened this issue: https://github.com/sequelize/sequelize/issues/8120 Do you think it can have anything to do with it?

**Comment by benjamn** [comment]
@molerat619 What version of Meteor are you using? Meteor 1.5.1 contains substantial improvements for this issue.

**Comment by molerat619** [comment]
@benjamn I'm not using Meteor at all but I've run the same commands and they solved my other issue I had back then. So I am asking now if that could have caused my new problem since it also has to do with sizes and nothing else makes sense to me.

**Comment by molerat619** [comment]
Never mind, I have found the problem and it was not caused by these commands. Sorry

**Comment by abernix** [comment]
Any new Meteor users experiencing this same problem should open a new issue.  Thanks!
